File: DSA/SakshamDwivedi_590013147/Day20/D20P3.py
import random

#Solution Function
def ranked_largest(arr, k):
    for i in range(len(arr)):
        kt = 1
        for j in range(len(arr)):
            if arr[i] < arr[j]:
                kt += 1
        if kt == k:
            return arr[i]
    return None


#Efficient solution using quickselect (Came through research)
def quickselect(arr, k):
    if not arr:
        return None
    
    sel = random.choice(arr)
    
    left = [x for x in arr if x > sel]
    mid = [x for x in arr if x == sel]
    right = [x for x in arr if x < sel]

    if k <= len(left):
        return quickselect(left, k)
    elif k <= len(left) + len(mid):
        return sel
    else:
        return quickselect(right, k - len(left) - len(mid))
    
#Fixed Input
arr = [3, 2, 1, 5, 6, 4]
k = 2
print(ranked_largest(arr, k))
print(quickselect(arr, k))

#Custom Input
arr = list(int(x) for x in input("Enter the array : ").split())
k = int(input("Enter the rank : "))
print(ranked_largest(arr, k))
print(quickselect(arr, k))

File: DSA/SakshamDwivedi_590013147/Day20/D20P3.c
#include <stdio.h>

int arr[100];
int size;
int k;
int i, j, kt;

int main() {
    printf("Enter the size of the array : ");
    scanf("%d", &size);
    printf("Enter the elements : ");
    for(i = 0; i < size; i++) {
        scanf("%d", &arr[i]);
    }
    printf("Enter the value of k : ");
    scanf("%d", &k);

    for(i = 0; i < size; i++) {
        kt = 1;
        for(j = 0; j < size; j++) {
            if(arr[i] < arr[j]) {
                kt++;
            }
        }
        if(kt == k) {
            printf("%d\n", arr[i]);
            return 0;
        }
    }

    printf("None\n");
}
